### ISSUE TITLE
Language Server verifying forever and/or leaving orphaned dotnet/z3 processes

### ISSUE THREAD

**Issue by hmijail** [dafny-lang/dafny#1611]
Dafny commit 1fc7bff2ccaf1b30297efc9bc2c6c9fa26b1a493
VSCode 1.62.3
Dafny Extension 2.0.2 using the Language Server in my Dafny build
macOS 11.6

If I start VSCode with this file open (related to the one in https://github.com/dafny-lang/dafny/issues/1570) :
```dafny
   newtype uint256 = i:nat | 0 <= i < 0x1_0000_0000_0000_0000_0000_0000_0000_0000_0000_0000_0000_0000_0000_0000_0000_0000  
   const MAX_UINT256 : uint256 := (0x1_0000_0000_0000_0000_0000_0000_0000_0000_0000_0000_0000_0000_0000_0000_0000_0000 as nat - 1) as uint256

   type uint = uint256 

   type bytes32 = uint256


method mul(a: uint256, b: uint256)

  {

    assert a as nat * b as nat >= 0;


    var c: nat := a as nat;
    var d: nat := b as nat;
    assert c * d >= 0;

    var a1: nat := a as nat;
    var b1: nat := b as nat; 
    
    //assert !(a1 * b1 < 0);
    //assert a1 * b1 >= 0;
    if (a1 >= 0 && b1 >= 0) {
          //assert a1 * b1 >= 0;

      assert !(a1 * b1 < 0); 
      // assert false;
    }
  }
```

The status bar shows "Verified" in a couple of seconds, but one (sometimes 2) z3 process(es) keep consuming 100% CPU for a time - some times a few seconds, some times tens of seconds, _even if I quit VSCode_. So either the "verified" status was reported too early, or Z3 kept running when it was no longer necessary?
At this point, even if Z3 finishes and disappears, a dotnet process seems to be left behind, using 0% CPU. Every time I open and close VSCode, a new dotnet process is leaked.

Relatedly, I have another, bigger file (related to this one) which causes the status bar to print "Verifying..." and stays like that forever, again with a Z3 process consuming 100% CPU, even if I quit VSCode. Should I try to reduce that file or is there any other way to help debug this?

**Comment by MikaelMayer** [comment]
I believe that the extension is running almost on every keystroke and previous spawned processes like Z3 are not killed, but they are solving problems that are not relevant anymore.
 @camrein  do you confirm?

**Comment by hmijail** [comment]
Note that I was describing what happens when opening VSCode with the file already open - no keystrokes involved. 

With this context, I expected one z3 process to run until the file is decided to be verified or not. What is surprising to me (not being any expert!) is that there are multiple z3 processes, and that they survive not only the apparent end of the verification process, but even VSCode quitting.

**Comment by camrein** [comment]
> Note that I was describing what happens when opening VSCode with the file already open - no keystrokes involved.
> 
> With this context, I expected one z3 process to run until the file is decided to be verified or not. What is surprising to me (not being any expert!) is that there are multiple z3 processes, and that they survive not only the apparent end of the verification process, but even VSCode quitting.

You're correct. The language server runs the automatic verification upon every document change (`onchange` setting in VSCode). The verification also happens when opening a document. Nevertheless, a change to the document should stop any pending verification. But we have observed that this doesn't work reliably, hence why #1493 was opened. Usually, the cancellation of boogie verification works, but sometimes not and I couldn't find out why yet.

> Note that I was describing what happens when opening VSCode with the file already open - no keystrokes involved.

As mentioned, automatic verification also runs when opening a document if set to `onchange`.


> With this context, I expected one z3 process to run until the file is decided to be verified or not. What is surprising to me (not being any expert!) is that there are multiple z3 processes, and that they survive not only the apparent end of the verification process, but even VSCode quitting.

The language server now uses half the cores for verification with the most recent release (see #1431). Therefore, you may observe more than one process (one process per core, to be more precise). Interestingly, the processes survive after their parent process is killed (VSCode and the language server). We may need to have a look at boogie how it spawns the Z3 processes, i.e., if they're deamons.

**Comment by hmijail** [comment]
Tangentially related: when the neverending "Verifying" is shown with its circular arrow spinning, 2 VSCode processes consume a combined 12-25% CPU (plus probably some GPU?). The processes are named "Code Helper (Renderer)" and "Code Helper (GPU)".
Switching to another tab stops showing the animation and those processes go back to ~0%. 

Probably this is a relatively minor problem, but still turnis this kind of bug into an extra pain point - I can't stop the animation affecting my battery life.

Should I report this elsewhere?

**Comment by camrein** [comment]
> Tangentially related: when the neverending "Verifying" is shown with its circular arrow spinning, 2 VSCode processes consume a combined 12-25% CPU (plus probably some GPU?). The processes are named "Code Helper (Renderer)" and "Code Helper (GPU)".
> Switching to another tab stops showing the animation and those processes go back to ~0%.

It's unlikely that the high load is caused by the extension itself. The extension merely displays information provided by the language server. In this case, that is the start of a verification. The extension only updates the status bar with a new icon and instructs VSCode to spin it using a standardized API: `$(sync~spin) Verifying...` The `~spin` part instructs VSCode to spin the icon `sync`.
What I'm wondering now: Does the high CPU load return when you switch back to the problematic tab? And does the status bar show `Verifying...` again?

> Probably this is a relatively minor problem, but still turnis this kind of bug into an extra pain point - I can't stop the animation affecting my battery life.

I agree that this shouldn't happen. Of course, the verification itself is CPU intensive. But that simple animation shouldn't be a problem.

> Should I report this elsewhere?

I have the feeling that reporting this to VSCode will lead them back to our extension as of now.  We probably have to strip it down to a minimal implementation that reproduces the CPU load. Does VSCode consume always this many resources when verifying?

**Comment by hmijail** [comment]
> Does the high CPU load return when you switch back to the problematic tab? And does the status bar show Verifying... again?

Yes to both.

> Does VSCode consume always this many resources when verifying?

Usually, while verifying, other processes are having much higher and longer spikes (z3, dotnet). And when those finish, then the verification and its animation also finish. So I guess that the big spikes always buried this measly 25% in the noise. 

I'm only realizing now because the animation runs without z3 and friends hiding it, plus it never stops. (By the way, I'm missing the "restart language server" option!)

**Comment by camrein** [comment]
> I'm only realizing now because the animation runs without z3 and friends hiding it, plus it never stops.

Unfortunately, I couldn't reproduce this issue on my Ubuntu and Windows machines yet (I do no longer have an OSX device at hand). But judging from your description, it almost has to be related to the spinning animation since the extension itself does nothing besides updating the message when switching tabs.


> By the way, I'm missing the "restart language server" option!

Could you elaborate on the reason for this (feel free to open an issue in https://github.com/dafny-lang/ide-vscode)? I didn't include it in the new revision since the only reason I could think of is when the language server hangs/crashes. These cases are probably bugs within the language server and should be fixed. Thinking of it, it might also be necessary if the verification is running (almost) infinitely and can't be canceled. A running verification should be cancelable, but it's not always working as stated in #1489.

**Comment by hmijail** [comment]
I guess/hope that both these issues (CPU/GPU usage by the animation, need of an option to restart the LS) won't matter anymore once the LS is working correctly (verification and its animation ending on time). So I'll just wait for the LS to be fixed :)

**Comment by MikaelMayer** [comment]
This fix might be related to the LS problem:
https://github.com/dafny-lang/ide-vscode/pull/93

**Comment by keyboardDrummer** [comment]
Orphaned processes should be resolved by https://github.com/dafny-lang/dafny/pull/1664

**Comment by atomb** [comment]
And displaying "Verifying..." forever when the server and Z3 _aren't_ actually doing anything should be resolved by #1652. That could be part of this issue, but not in the case where either the language server or Z3 are busy.
